File: python-version/fingerprinting/algorithm.py
#!/usr/bin/python3
#-*- encoding: Utf-8 -*-
from numpy import fft, array as nparray, maximum, log, hanning, mean, abs, round
from typing import Dict, List, Set, Sequence, Union, Optional, Any
from struct import pack, unpack
from enum import IntEnum
from copy import copy

HANNING_MATRIX = hanning(2050)[1:-1] # Wipe trailing and leading zeroes


from signature_format import DecodedMessage, FrequencyPeak, RawSignatureHeader, FrequencyBand


class RingBuffer(list):
    
    def __init__(self, buffer_size : int, default_value : Any = None):
        
        if default_value is not None:
            list.__init__(self, [copy(default_value) for position in range(buffer_size)])
        else:
            list.__init__(self, [None] * buffer_size)
        
        self.position : int = 0
        self.buffer_size : int = buffer_size
        self.num_written : int = 0
    
    def append(self, value : Any):
        
        self[self.position] = value
        
        self.position += 1
        self.position %= self.buffer_size
        self.num_written += 1
        
class SignatureGenerator:
    
    def __init__(self):
        
        # Used when storing input that will be processed when requiring to
        # generate a signature:
        
        self.input_pending_processing : List[int] = [] # Signed 16-bits, 16 KHz mono samples to be processed
        
        self.samples_processed : int = 0 # Number of samples processed out of "self.input_pending_processing"
        
        # Used when processing input:
        
        self.ring_buffer_of_samples : RingBuffer[int] = RingBuffer(buffer_size = 2048, default_value = 0)
        
        self.fft_outputs : RingBuffer[List[float]] = RingBuffer(buffer_size = 256, default_value = [0. * 1025]) # Lists of 1025 floats, premultiplied with a Hanning function before being passed through FFT, computed from the ring buffer every new 128 samples
        
        self.spread_ffts_output : RingBuffer[List[float]] = RingBuffer(buffer_size = 256, default_value = [0] * 1025)

        # How much data to send to Shazam at once?

        self.MAX_TIME_SECONDS = 3.1
        self.MAX_PEAKS = 255
        
        # The object that will hold information about the next fingerpring
        # to be produced
        
        self.next_signature = DecodedMessage()
        self.next_signature.sample_rate_hz = 16000
        self.next_signature.number_samples = 0
        self.next_signature.frequency_band_to_sound_peaks = {}
    
    """
        Add data to be generated a signature for, which will be
        processed when self.get_next_signature() is called. This
        function expects signed 16-bit 16 KHz mono PCM samples.
    """
    
    def feed_input(self, s16le_mono_samples : List[int]):
        
        self.input_pending_processing += s16le_mono_samples
    
    """
        Consume some of the samples fed to self.feed_input(), and return
        a Shazam signature (DecodedMessage object) to be sent to servers
        once "enough data has been gathered".
        
        Except if there are no more samples to be consumed, in this case
        we will return None.
    """
    
    def get_next_signature(self) -> Optional[DecodedMessage]:
        
        if len(self.input_pending_processing) - self.samples_processed < 128:
            return None
        
        while (len(self.input_pending_processing) - self.samples_processed >= 128 and
            (self.next_signature.number_samples / self.next_signature.sample_rate_hz < self.MAX_TIME_SECONDS or
            sum(len(peaks) for peaks in self.next_signature.frequency_band_to_sound_peaks.values()) < self.MAX_PEAKS
            )):
            
            self.process_input(self.input_pending_processing[self.samples_processed:self.samples_processed + 128])
            
            self.samples_processed += 128

        returned_signature = self.next_signature

        self.next_signature = DecodedMessage()
        self.next_signature.sample_rate_hz = 16000
        self.next_signature.number_samples = 0
        self.next_signature.frequency_band_to_sound_peaks = {}
        
        self.ring_buffer_of_samples : RingBuffer[int] = RingBuffer(buffer_size = 2048, default_value = 0)
        self.fft_outputs : RingBuffer[List[float]] = RingBuffer(buffer_size = 256, default_value = [0. * 1025])
        self.spread_ffts_output : RingBuffer[List[float]] = RingBuffer(buffer_size = 256, default_value = [0] * 1025)
        
        return returned_signature

    
    def process_input(self, s16le_mono_samples : List[int]):
    
        self.next_signature.number_samples += len(s16le_mono_samples)
        
        for position_of_chunk in range(0, len(s16le_mono_samples), 128):
            
            self.do_fft(s16le_mono_samples[position_of_chunk:position_of_chunk + 128])
            
            self.do_peak_spreading_and_recognition()
        
    def do_fft(self, batch_of_128_s16le_mono_samples):
        
        self.ring_buffer_of_samples[
            self.ring_buffer_of_samples.position:
            self.ring_buffer_of_samples.position + len(batch_of_128_s16le_mono_samples)
        ] = batch_of_128_s16le_mono_samples
        
        self.ring_buffer_of_samples.position += len(batch_of_128_s16le_mono_samples)
        self.ring_buffer_of_samples.position %= 2048
        self.ring_buffer_of_samples.num_written +=  len(batch_of_128_s16le_mono_samples)
        
        excerpt_from_ring_buffer : list = (
            self.ring_buffer_of_samples[self.ring_buffer_of_samples.position:] +
            self.ring_buffer_of_samples[:self.ring_buffer_of_samples.position]
        )
        
        # The premultiplication of the array is for applying a windowing function before the DFT (slighty rounded Hanning without zeros at edges)
        
        fft_results : nparray = fft.rfft(HANNING_MATRIX * excerpt_from_ring_buffer)

        assert len(fft_results) == 1025 and len(excerpt_from_ring_buffer) == 2048 == len(HANNING_MATRIX)
        
        fft_results = (fft_results.real ** 2 + fft_results.imag ** 2) / (1 << 17)
        fft_results = maximum(fft_results, 0.0000000001)
        
        self.fft_outputs.append(fft_results)
        
    
    def do_peak_spreading_and_recognition(self):
        
        self.do_peak_spreading()
        
        if self.spread_ffts_output.num_written >= 46:
            
            self.do_peak_recognition()
    
    def do_peak_spreading(self):
        
        origin_last_fft : List[float] = self.fft_outputs[self.fft_outputs.position - 1]
        
        spread_last_fft : List[float] = list(origin_last_fft)
        
        for position in range(1025):
            
            # Perform frequency-domain spreading of peak values
            
            if position < 1023:
            
                spread_last_fft[position] = max(spread_last_fft[position:position + 3])
            
            # Perform time-domain spreading of peak values
            
            max_value = spread_last_fft[position]
            
            for former_fft_num in [-1, -3, -6]:
                
                former_fft_output = self.spread_ffts_output[(self.spread_ffts_output.position + former_fft_num) % self.spread_ffts_output.buffer_size]
                
                former_fft_output[position] = max_value = max(former_fft_output[position], max_value)
                
        # Save output locally
        
        self.spread_ffts_output.append(spread_last_fft)
        
        pass
    
    def do_peak_recognition(self):
        
        fft_minus_46 = self.fft_outputs[(self.fft_outputs.position - 46) % self.fft_outputs.buffer_size]
        fft_minus_49 = self.spread_ffts_output[(self.spread_ffts_output.position - 49) % self.spread_ffts_output.buffer_size]
        fft_minus_53 = self.spread_ffts_output[(self.spread_ffts_output.position - 53) % self.spread_ffts_output.buffer_size]
        fft_minus_45 = self.spread_ffts_output[(self.spread_ffts_output.position - 45) % self.spread_ffts_output.buffer_size]
        
        for bin_position in range(10, 1015):
            
            # Ensure that the bin is large enough to be a peak
            
            if (fft_minus_46[bin_position] >= 1 / 64 and
                fft_minus_46[bin_position] >= fft_minus_49[bin_position - 1]):
                
                # Ensure that it is frequency-domain local minimum
                
                max_neighbor_in_fft_minus_49 = 0
                
                for neighbor_offset in [*range(-10, -3, 3), -3, 1, *range(2, 9, 3)]:
                
                    max_neighbor_in_fft_minus_49 = max(fft_minus_49[bin_position + neighbor_offset], max_neighbor_in_fft_minus_49)
                
                if fft_minus_46[bin_position] > max_neighbor_in_fft_minus_49:
                    
                    # Ensure that it is a time-domain local minimum
                    
                    max_neighbor_in_other_adjacent_ffts = max_neighbor_in_fft_minus_49
                    
                    for other_offset in [-53, -45, *range(165, 201, 7), *range(214, 250, 7)]:
                    
                        max_neighbor_in_other_adjacent_ffts = max(
                            self.spread_ffts_output[(self.spread_ffts_output.position + other_offset) % self.spread_ffts_output.buffer_size][bin_position - 1],
                            max_neighbor_in_other_adjacent_ffts
                        )
                    
                    if fft_minus_46[bin_position] > max_neighbor_in_other_adjacent_ffts:
                        
                        # This is a peak, store the peak
                        
                        fft_number = self.spread_ffts_output.num_written - 46
                        
                        peak_magnitude = log(max(1 / 64, fft_minus_46[bin_position])) * 1477.3 + 6144
                        peak_magnitude_before = log(max(1 / 64, fft_minus_46[bin_position - 1])) * 1477.3 + 6144
                        peak_magnitude_after = log(max(1 / 64, fft_minus_46[bin_position + 1])) * 1477.3 + 6144
                        
                        peak_variation_1 = peak_magnitude * 2 - peak_magnitude_before - peak_magnitude_after
                        peak_variation_2 = (peak_magnitude_after - peak_magnitude_before) * 32 / peak_variation_1
                        
                        corrected_peak_frequency_bin = bin_position * 64 + peak_variation_2
                        
                        assert peak_variation_1 > 0
                        
                        frequency_hz = corrected_peak_frequency_bin * (16000 / 2 / 1024 / 64)
                        
                        if frequency_hz < 250:
                            continue
                        elif frequency_hz < 520:
                            band = FrequencyBand._250_520
                        elif frequency_hz < 1450:
                            band = FrequencyBand._520_1450
                        elif frequency_hz < 3500:
                            band = FrequencyBand._1450_3500
                        elif frequency_hz <= 5500:
                            band = FrequencyBand._3500_5500
                        else:
                            continue
                        
                        if band not in self.next_signature.frequency_band_to_sound_peaks:
                            self.next_signature.frequency_band_to_sound_peaks[band] = []
                        
                        self.next_signature.frequency_band_to_sound_peaks[band].append(
                            FrequencyPeak(fft_number, int(peak_magnitude), int(corrected_peak_frequency_bin), 16000)
                        )
                        

File: src/fingerprinting/algorithm.rs
use chfft::RFft1D;
use std::error::Error;
use std::io::BufReader;
use std::collections::HashMap;

use crate::fingerprinting::hanning::HANNING_WINDOW_2048_MULTIPLIERS;
use crate::fingerprinting::signature_format::{DecodedSignature, FrequencyBand, FrequencyPeak};
#[cfg(feature = "ffmpeg")]
use crate::utils::ffmpeg_wrapper::decode_with_ffmpeg;


pub struct SignatureGenerator {

    // Used when processing input:

    ring_buffer_of_samples: Vec<i16>,
    /// Ring buffer.
    ring_buffer_of_samples_index: usize,

    reordered_ring_buffer_of_samples: Vec<f32>,
    /// Reordered, temporary version of the ring buffer above, with floats for precision because we applied Hanning window.

    fft_outputs: Vec<Vec<f32>>,
    /// Ring buffer. Lists of 1025 floats, premultiplied with a Hanning function before being passed through FFT, computed from the ring buffer every new 128 samples
    fft_outputs_index: usize,

    fft_object: RFft1D<f32>,

    spread_fft_outputs: Vec<Vec<f32>>,
    /// Ring buffer.
    spread_fft_outputs_index: usize,

    num_spread_ffts_done: u32,

    signature: DecodedSignature,
}

impl SignatureGenerator {
    pub fn make_signature_from_file(file_path: &str) -> Result<DecodedSignature, Box<dyn Error>> {

        // Decode the .WAV, .MP3, .OGG or .FLAC file
        
        #[cfg(not(feature = "ffmpeg"))]
        let decoder = rodio::Decoder::new(BufReader::new(std::fs::File::open(file_path)?));

        #[cfg(feature = "ffmpeg")]
        let decoder = {
            let mut decoder = rodio::Decoder::new(BufReader::new(std::fs::File::open(file_path)?));

            if let Err(ref _decoding_error) = decoder {
                
                // Try to decode with FFMpeg, if available, in case of failure with
                // Rodio (most likely due to the use of a format unsupported by
                // Rodio, such as .WMA or .MP4/.AAC)
                
                if let Some(new_decoder) = decode_with_ffmpeg(file_path) {
                    decoder = Ok(new_decoder);
                }
            }

            decoder
        };
        
        // Downsample the raw PCM samples to 16 KHz, and skip to the middle of the file
        // in order to increase recognition odds. Take 12 seconds of sample.

        let converted_file = rodio::source::UniformSourceIterator::new(decoder?, 1, 16000);

        let raw_pcm_samples: Vec<i16> = converted_file.collect();
        let mut raw_pcm_samples_slice: &[i16] = &raw_pcm_samples;

        let slice_len = raw_pcm_samples_slice.len().min(12 * 16000);

        if raw_pcm_samples_slice.len() > 12 * 16000 {
            let middle = raw_pcm_samples.len() / 2;

            raw_pcm_samples_slice = &raw_pcm_samples_slice[middle - (6 * 16000)..middle + (6 * 16000)];
        }

        Ok(SignatureGenerator::make_signature_from_buffer(&raw_pcm_samples_slice[..slice_len]))
    }

    pub fn make_signature_from_buffer(s16_mono_16khz_buffer: &[i16]) -> DecodedSignature {
        let mut this = SignatureGenerator {
            ring_buffer_of_samples: vec![0i16; 2048],
            ring_buffer_of_samples_index: 0,

            reordered_ring_buffer_of_samples: vec![0.0f32; 2048],

            fft_outputs: vec![vec![0.0f32; 1025]; 256],
            fft_outputs_index: 0,

            fft_object: RFft1D::<f32>::new(2048),

            spread_fft_outputs: vec![vec![0.0f32; 1025]; 256],
            spread_fft_outputs_index: 0,

            num_spread_ffts_done: 0,

            signature: DecodedSignature {
                sample_rate_hz: 16000,
                number_samples: s16_mono_16khz_buffer.len() as u32,
                frequency_band_to_sound_peaks: HashMap::new(),
            },
        };

        for chunk in s16_mono_16khz_buffer.chunks_exact(128) {
            this.do_fft(chunk);

            this.do_peak_spreading();

            this.num_spread_ffts_done += 1;

            if this.num_spread_ffts_done >= 46 {
                this.do_peak_recognition();
            }
        }

        this.signature
    }

    fn do_fft(self: &mut Self, s16_mono_16khz_buffer: &[i16]) {

        // Copy the 128 input s16le samples to the local ring buffer

        self.ring_buffer_of_samples[self.ring_buffer_of_samples_index..self.ring_buffer_of_samples_index + 128].copy_from_slice(s16_mono_16khz_buffer);

        self.ring_buffer_of_samples_index += 128;
        self.ring_buffer_of_samples_index &= 2047;

        // Reorder the items (put the latest data at end) and apply Hanning window

        for index in 0..=2047 {
            self.reordered_ring_buffer_of_samples[index] =
                self.ring_buffer_of_samples[(index + self.ring_buffer_of_samples_index) & 2047] as f32 *
                    HANNING_WINDOW_2048_MULTIPLIERS[index];
        }

        // Perform Fast Fourier transform

        let complex_fft_results = self.fft_object.forward(&self.reordered_ring_buffer_of_samples);

        assert_eq!(complex_fft_results.len(), 1025);

        // Turn complex into reals, and put the results into a local array

        let real_fft_results = &mut self.fft_outputs[self.fft_outputs_index];

        for index in 0..=1024 {
            real_fft_results[index] = (
                (
                    complex_fft_results[index].re.powi(2) +
                        complex_fft_results[index].im.powi(2)
                ) / ((1 << 17) as f32)
            ).max(0.0000000001);
        }

        self.fft_outputs_index += 1;
        self.fft_outputs_index &= 255;
    }

    fn do_peak_spreading(self: &mut Self) {
        let real_fft_results = &self.fft_outputs[((self.fft_outputs_index as i32 - 1) & 255) as usize];

        let spread_fft_results = &mut self.spread_fft_outputs[self.spread_fft_outputs_index];

        // Perform frequency-domain spreading of peak values

        spread_fft_results.copy_from_slice(real_fft_results);

        for position in 0..=1022 {
            spread_fft_results[position] = spread_fft_results[position]
                .max(spread_fft_results[position + 1])
                .max(spread_fft_results[position + 2]);
        }

        // Perform time-domain spreading of peak values

        let spread_fft_results_copy = spread_fft_results.clone(); // Avoid mutable+mutable borrow of self.spread_fft_outputs

        for position in 0..=1024 {
            for former_fft_number in &[1, 3, 6] {
                let former_fft_output = &mut self.spread_fft_outputs[((self.spread_fft_outputs_index as i32 - *former_fft_number) & 255) as usize];

                former_fft_output[position] = former_fft_output[position]
                    .max(spread_fft_results_copy[position]);
            }
        }

        self.spread_fft_outputs_index += 1;
        self.spread_fft_outputs_index &= 255;
    }

    fn do_peak_recognition(self: &mut Self) {

        // Note: when substracting an array index, casting to signed is needed
        // to avoid underflow panics at runtime.

        let fft_minus_46 = &self.fft_outputs[((self.fft_outputs_index as i32 - 46) & 255) as usize];
        let fft_minus_49 = &self.spread_fft_outputs[((self.spread_fft_outputs_index as i32 - 49) & 255) as usize];

        for bin_position in 10..=1014 {

            // Ensure that the bin is large enough to be a peak

            if fft_minus_46[bin_position] >= 1.0 / 64.0 &&
                fft_minus_46[bin_position] >= fft_minus_49[bin_position - 1] {

                // Ensure that it is frequency-domain local minimum

                let mut max_neighbor_in_fft_minus_49: f32 = 0.0;

                for neighbor_offset in &[-10, -7, -4, -3, 1, 2, 5, 8] {
                    max_neighbor_in_fft_minus_49 = max_neighbor_in_fft_minus_49
                        .max(fft_minus_49[(bin_position as i32 + *neighbor_offset) as usize]);
                }

                if fft_minus_46[bin_position] > max_neighbor_in_fft_minus_49 {

                    // Ensure that it is a time-domain local minimum

                    let mut max_neighbor_in_other_adjacent_ffts = max_neighbor_in_fft_minus_49;

                    for other_offset in &[-53, -45,
                        165, 172, 179, 186, 193, 200,
                        214, 221, 228, 235, 242, 249] {
                        let other_fft = &self.spread_fft_outputs[((self.spread_fft_outputs_index as i32 + other_offset) & 255) as usize];

                        max_neighbor_in_other_adjacent_ffts = max_neighbor_in_other_adjacent_ffts
                            .max(other_fft[bin_position - 1]);
                    }

                    if fft_minus_46[bin_position] > max_neighbor_in_other_adjacent_ffts {

                        // This is a peak, store the peak

                        let fft_pass_number = self.num_spread_ffts_done - 46;

                        let peak_magnitude: f32 = fft_minus_46[bin_position].ln().max(1.0 / 64.0) * 1477.3 + 6144.0;
                        let peak_magnitude_before: f32 = fft_minus_46[bin_position - 1].ln().max(1.0 / 64.0) * 1477.3 + 6144.0;
                        let peak_magnitude_after: f32 = fft_minus_46[bin_position + 1].ln().max(1.0 / 64.0) * 1477.3 + 6144.0;

                        let peak_variation_1: f32 = peak_magnitude * 2.0 - peak_magnitude_before - peak_magnitude_after;
                        let peak_variation_2: f32 = (peak_magnitude_after - peak_magnitude_before) * 32.0 / peak_variation_1;

                        let corrected_peak_frequency_bin: u16 = bin_position as u16 * 64 + (peak_variation_2 as i16) as u16;

                        assert!(peak_variation_1 >= 0.0);

                        // Convert back a FFT bin to a frequency, given a 16 KHz sample
                        // rate, 1024 useful bins and the multiplication by 64 made before
                        // storing the information

                        let frequency_hz: f32 = corrected_peak_frequency_bin as f32 * (16000.0 / 2.0 / 1024.0 / 64.0);

                        // Ignore peaks outside the 250 Hz-5.5 KHz range, store them into
                        // a lookup table that will be used to generate the binary fingerprint
                        // otherwise

                        let frequency_band = match frequency_hz as i32 {
                            250..=519 => FrequencyBand::_250_520,
                            520..=1449 => FrequencyBand::_520_1450,
                            1450..=3499 => FrequencyBand::_1450_3500,
                            3500..=5500 => FrequencyBand::_3500_5500,
                            _ => { continue; }
                        };

                        if !self.signature.frequency_band_to_sound_peaks.contains_key(&frequency_band) {
                            self.signature.frequency_band_to_sound_peaks.insert(frequency_band, vec![]);
                        }

                        self.signature.frequency_band_to_sound_peaks.get_mut(&frequency_band).unwrap().push(
                            FrequencyPeak {
                                fft_pass_number: fft_pass_number,
                                peak_magnitude: peak_magnitude as u16,
                                corrected_peak_frequency_bin: corrected_peak_frequency_bin,
                                sample_rate_hz: 16000,
                            }
                        );
                    }
                }
            }
        }
    }
}
